fbw-common/src/wasm/systems/systems/src/shared/logic_nodes.rs

```diff
@@ -0,0 +1,624 @@
+        self.output
+    }
+}
+
+/// A flip-flop or memory circuit that can be used to store a single bit. It has two inputs: Set and
+/// Reset. At first it will always emit a falsy value, until it receives a signal on the set input,
+/// at which point it will start emitting a truthy value. This will continue until a signal is
+/// received on the reset input, at which point it reverts to the original falsy output. If a signal
+/// is received on both set and reset at the same time, the input with a star will have precedence.
+pub struct MemoryNode {
+    has_set_precedence: bool,
+    pub output: bool,
+}
+
+impl MemoryNode {
+    pub fn new(has_set_precedence: bool) -> Self {
+        Self {
+            has_set_precedence,
+            output: false,
+        }
+    }
+
+    pub fn update(&mut self, set: bool, reset: bool) -> bool {
+        self.output = if set && reset {
+            self.has_set_precedence
+        } else if set {
+            true
+        } else if reset {
+            false
+        } else {
+            self.output
+        };
+        self.output
+    }
+}
+
+/// A hysteresis circuit, which will switch between a high and a low state based on two different
+/// numerical comparisons to prevent rapid output switching due to minor value fluctuations.
+/// The circuit will output lo until the up condition is met (value >= up). Then, even if the
+/// condition subsequently fails, the circuit will continue outputting hi until the down condition
+/// is met (value <= dn).
+pub struct HysteresisNode<T> {
+    up: T,
+    dn: T,
+    output: bool,
+}
+
+impl<T> HysteresisNode<T>
+where
+    T: PartialOrd,
+{
+    pub fn new(dn: T, up: T) -> Self {
+        if up <= dn {
+            panic!("Up must be strictly greater than Dn");
+        }
+        Self {
+            up,
+            dn,
+            output: false,
+        }
+    }
+
+    pub fn update(&mut self, value: T) -> bool {
+        self.output = if self.output {
+            if value <= self.dn {
+                false
+            } else {
+                self.output
+            }
+        } else if value >= self.up {
+            true
+        } else {
+            self.output
+        };
+        self.output
+    }
+}
+
+/// A node that memorizes the value from the preceding call.
+#[derive(Default)]
+pub struct PreceedingValueNode {
+    predecessor: bool,
+}
+
+impl PreceedingValueNode {
+    pub fn new() -> Self {
+        Self { predecessor: false }
+    }
+
+    pub fn value(&self) -> bool {
+        self.predecessor
+    }
+
+    pub fn update(&mut self, value: bool) {
+        self.predecessor = value;
+    }
+}
+
+/// A node that compares the derivative of a value over time to a threshold and returns true if the
+/// derivative is strictly larger. The threshold is interpreted as change per 1 second.
+/// The derivative is calculated over two subsequent .update calls.
+#[derive(Default)]
+pub struct DerivativeThresholdNode {
```

Comment:
Can this node not be implemented with the `DerivativeNode` from `derivative.rs`? Would reduce code duplication.